compiler/rustc_middle/src/ty/context.rs

```diff
@@ -595,10 +600,60 @@ impl<'tcx> GlobalCtxt<'tcx> {
         F: FnOnce(TyCtxt<'tcx>) -> R,
     {
         let icx = tls::ImplicitCtxt::new(self);
+
+        let gcx_ptr =
+            GcxPtr { value: Lrc::new(RwLock::new(Some(icx.tcx.gcx as *const _ as *const ()))) };
+
+        // Reset `gcx_ptr` to `None` when we exit.
+        let gcx_ptr_ = gcx_ptr.clone();
+        let _on_drop = defer(move || {
+            *gcx_ptr_.value.write() = None;
+        });
+
+        // Set this `GlobalCtxt` as the current one.
+        *self.current_gcx.value.lock() = Some(gcx_ptr);
+
         tls::enter_context(&icx, || f(icx.tcx))
     }
 }
 
+/// This stores a pointer to a `GlobalCtxt`. When the `GlobalCtxt` is no longer available the lock
+/// will be set to `None`.
+#[derive(Clone)]
+struct GcxPtr {
+    value: Lrc<RwLock<Option<*const ()>>>,
+}
+
+#[cfg(parallel_compiler)]
+unsafe impl DynSend for GcxPtr {}
+
+/// This is used to get a reference to a `GlobalCtxt` if one is available.
+///
+/// This is needed to allow the deadlock handler access to `GlobalCtxt` to look for query cycles.
+/// It cannot use the `TLV` global because that's only guaranteed to be defined on the thread
+/// creating the `GlobalCtxt`. Other threads have access to the `TLV` only inside Rayon jobs, but
+/// the deadlock handler is not called inside such a job.
+#[derive(Clone)]
+pub struct CurrentGcx {
+    value: Lrc<Lock<Option<GcxPtr>>>,
```

Comment:
`GcxPtr` contains a `Lrc<RwLock<Option<*const ()>>>`. Which means that `CurrentGcx` is basically a `Lrc<Lock<Option<Lrc<RwLock<Option<*const ()>>>>>>`.

I think this is the most complex Rust type I've ever seen. Something is very wrong here. Can it be made simpler?

Reply:
I was trying to make `enter` behave well when called concurrently, but I think `CurrentGcx` would have to hold a set of `GcxPtr` for that to work properly. I've removed `GcxPtr` and instead let `enter` panic for concurrent usage.